vault/acl.go

```diff
@@ -334,13 +367,81 @@ func (a *ACL) AllowOperation(ctx context.Context, req *logical.Request, capCheck
 		}
 	}
 
-	// Find a glob rule, default deny if no match
+	// Find a prefix rule, default deny if no match
 	_, raw, ok = a.prefixRules.LongestPrefix(path)
-	if !ok {
-		return
+	if ok {
+		permissions = raw.(*ACLPermissions)
+		capabilities = permissions.CapabilitiesBitmap
+		goto CHECK
 	}
-	permissions = raw.(*ACLPermissions)
-	capabilities = permissions.CapabilitiesBitmap
+
+	if len(a.segmentWildcardPaths) > 0 {
+		pathParts := strings.Split(path, "/")
+		for currWCPath := range a.segmentWildcardPaths {
+			if currWCPath == "" {
+				continue
+			}
+
+			var isPrefix bool
+			var invalid bool
+			origCurrWCPath := currWCPath
+
+			if currWCPath[len(currWCPath)-1] == '*' {
+				isPrefix = true
+				currWCPath = currWCPath[0 : len(currWCPath)-1]
+			}
+			splitCurrWCPath := strings.Split(currWCPath, "/")
+			if len(pathParts) < len(splitCurrWCPath) {
+				// The path coming in is shorter; it can't match
+				continue
+			}
+			if !isPrefix && len(splitCurrWCPath) != len(pathParts) {
+				// If it's not a prefix we expect the same number of segments
+				continue
+			}
+			// We key off splitK here since it might be less than pathParts
+			for i, aclPart := range splitCurrWCPath {
+				if aclPart == "+" {
+					// Matches anything in the segment, so keep checking
+					continue
+				}
+				if i == len(splitCurrWCPath)-1 && isPrefix {
+					// In this case we may have foo* or just * depending on if
+					// originally it was foo* or foo/*.
+					if aclPart == "" {
+						// Ended in /*, so at this point we're at the final
+						// glob which will match anything, so return success
+						break
+					}
+					if !strings.HasPrefix(pathParts[i], aclPart) {
+						// E.g., the final part of the acl is foo* and the
+						// final part of the path is boofar
+						invalid = true
+						break
+					}
+					// Final prefixed matched and the rest is a wildcard,
+					// matches
+					break
+				}
+				if aclPart != pathParts[i] {
+					// Mismatch, exit out
+					invalid = true
+					break
+				}
+			}
+			// If invalid isn't set then we got through the full segmented path
+			// without finding a mismatch, so it's valid
+			if !invalid {
```

Comment:
Might be safer to inverse this and set-if-valid. That way it's more explicit if any more conditions are added above

Reply:
The reason this is tough to do the other way is that exiting the for loop naturally (by running through the full range) is actually a success condition. If you finish the for loop you don't (easily) know if it's by breaking out or finishing it naturally.